addons/account/models/account_partial_reconcile.py

```diff
@@ -138,44 +138,42 @@ def create(self, vals_list):
-                                ELSE partial.debit_move_id
-                           END
-                      FROM partials p
-                      JOIN account_partial_reconcile partial ON p.current_id = partial.debit_move_id
-                                                             OR p.current_id = partial.credit_move_id
-            )
-              SELECT line_id, 'P' || MIN(partial.id) AS matching_number
-                FROM partials
-                JOIN account_partial_reconcile partial ON current_id = partial.debit_move_id
-                                                       OR current_id = partial.credit_move_id
-            GROUP BY line_id
-        """, [
-            amls.ids,
-        ])
-
-        line_matching_number = dict(self.env.cr.fetchall())
-
-        for line in amls.with_context(skip_invoice_sync=True):
-            if line.full_reconcile_id:
-                line.matching_number = str(line.full_reconcile_id.id)
-            elif line.matched_debit_ids or line.matched_credit_ids:
-                line.matching_number = line_matching_number[line.id]
+        all_partials = amls.matched_debit_ids | amls.matched_credit_ids
+
+        number2lines = {}
+        line2number = {}
+        for partial in all_partials.sorted('id'):
+            debit_min_id = line2number.get(partial.debit_move_id.id)
+            credit_min_id = line2number.get(partial.credit_move_id.id)
+            if debit_min_id and credit_min_id:
+                if debit_min_id != credit_min_id:
+                    min_min_id = min(debit_min_id, credit_min_id)
+                    max_min_id = max(debit_min_id, credit_min_id)
+                    for line_id in number2lines[max_min_id]:
+                        line2number[line_id] = min_min_id
+                    number2lines[min_min_id].extend(number2lines.pop(max_min_id))
+            elif debit_min_id:
+                number2lines[debit_min_id].append(partial.credit_move_id.id)
+                line2number[partial.credit_move_id.id] = debit_min_id
+            elif credit_min_id:
+                number2lines[credit_min_id].append(partial.debit_move_id.id)
+                line2number[partial.debit_move_id.id] = credit_min_id
             else:
-                line.matching_number = False
-
+                number2lines[partial.id] = [partial.debit_move_id.id, partial.credit_move_id.id]
+                line2number[partial.debit_move_id.id] = partial.id
+                line2number[partial.credit_move_id.id] = partial.id
+
+        processed_amls = []
+        with amls.move_id._check_balanced({'records': amls.move_id}):  # avoid checking the consistency for each individual write
+            for min_partial_id, line_ids in number2lines.items():
+                min_partial = self.browse(min_partial_id)
+                self.env['account.move.line'].browse(line_ids).matching_number = (
+                    str(min_partial.full_reconcile_id.id)
```

Comment:
At this point, this is not making a read for each partial thanks to the prefetch made when accessing `debit_move_id` earlier. I think an explicit `with_prefetch` will make the code a bit more understandable for a lambda developer.

Reply:
I think `with_prefetch` is confusing for lambda devs, they usually never have to worry about those things